File: 918.环形子数组的最大和.cpp
/*
 * @lc app=leetcode.cn id=918 lang=cpp
 *
 * [918] 环形子数组的最大和
 */

// @lc code=start
#include<vector>
#include<numeric>
using namespace std;
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        pair<int, int> dp_max = {nums[0], 1}, dp_min = {nums[0], 1};
        int ret = nums[0], min_val = nums[0], min_l = 1;
        for(int i=1; i<nums.size(); ++i){
            dp_max = max(pair<int, int>(dp_max.first + nums[i], dp_max.second + 1), pair<int, int>(nums[i], 1));
            dp_min = min(pair<int, int>(dp_min.first + nums[i], dp_min.second + 1), pair<int, int>(nums[i], 1));
        
            ret = max(ret, dp_max.first);
            if(dp_min.first < min_val) tie(min_val, min_l) = dp_min;
        }

        if(min_l < nums.size()) ret = max(ret, accumulate(nums.begin(), nums.end(), 0) - min_val);
        return ret;
    }
};
// @lc code=end

File: 918.环形子数组的最大和.py
#
# @lc app=leetcode.cn id=918 lang=python3
#
# [918] 环形子数组的最大和
#

# @lc code=start
from typing import List
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        dp_max, dp_min = (nums[0], 1), (nums[0], 1)
        ret, min_val, min_l = nums[0], nums[0], 1
        for i in range(1, len(nums)):
            dp_max = max((dp_max[0] + nums[i], dp_max[1] + 1), (nums[i], 1))
            dp_min = min((dp_min[0] + nums[i], dp_min[1] + 1), (nums[i], 1))
            
            ret = max(ret, dp_max[0])
            if dp_min[0] < min_val: min_val, min_l = dp_min

        if min_l < len(nums): ret = max(ret, sum(nums) - min_val)
        return ret
# @lc code=end

print(Solution().maxSubarraySumCircular(nums = [1,-2,3,-2]))
print(Solution().maxSubarraySumCircular(nums = [5,-3,5]))
print(Solution().maxSubarraySumCircular(nums = [3,-2,2,-3]))
print(Solution().maxSubarraySumCircular([-2]))